### PR TITLE
[feat] [FSDP]: add experimental support to shared weights

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?

This adds experimental and limited support of shared weights between FSDP wrapper.

### Why is this needed?

When a weight is shared between layers, before this change, it needs to be part of a single FSDP wrapper. This means that while the weight is not being used, it might still be stored in GPU memory in full due to this limitation.

When we can separately wrap layers with shared weights, GPU memory can be saved. E.g.

```
L0 -> L1 -> L2 -> L3
L3.weight = L0.weight
```

Without this change, L0 and L3's weight are in the outer (root) FSDP and can't be sharded when L1 and L2 are computing.

With this change, L0 and L3 can be separately wrapped. Thus, allow the weight and gradient to be sharded as soon as it is done computing.

A special type of L3 kernel will be needed due to autograd execution order. An example of that will be in a separate PR.

### Why is the limiting and experimental

The FSDP wrappers for the weight sharing layers need to be un-flatten and special care need to be taken with the layers. 

Supporting general sharing case is likely very difficult and not necessarily needed. This is to enable particular kind of big scale training for research purposes.

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
> @zhaojuanmao
> 
> Now this test's results make more sense.
> 
> * If nothing is shared, of course, FSDP and nonFSDP match.
> * If something is shared:
>   
>   1. shape mismatch error (non flat cases)
>   2. no shape mismatch but numeric mismatch (as if nothing is shared under FSDP) (flat cases)

I see, that makes sense to me as well!

**Comment by min-xu-ai** [comment]
@zhaojuanmao. I added some limited and experimental support for shared param. If you want to, please take a look!

**Review by sshleifer — COMMENTED** [review]
LGTM. Can we revert/keep the code simpler if this doesn't end up being useful?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -414,6 +414,33 @@ def module(self) -> FlattenParamsWrapper:
...
+                The shared parameter.
+        """
+        assert self._is_root is None
+        assert not self.flatten_parameters
+        assert isinstance(p, Parameter)
+        assert p._is_sharded
+        p._is_shared = True
+        assert (
+            len(list(filter(lambda p: not (hasattr(p, "_is_shared") and p._is_shared), self.params))) > 0
+        ), "Must have at least 1 non-shared param."
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:441]: what does this requirement satisfy?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1367,6 +1402,17 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
         if param.grad is None:
             return
 
+        if hasattr(param, "_linked_param"):
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1405]: where do we set this property?

File: tests/nn/data_parallel/test_fsdp_shared_weights.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        for _ in range(steps_per_iter):
+            out = model(in_data)
+            out.sum().backward()
+        # Simulate gradient means between ranks.
+        if steps_per_iter > 1:
+            with torch.no_grad():
+                for p in model.parameters():
+                    p.grad /= steps_per_iter
+        optim.step()
+        model.zero_grad(set_to_none=True)
```
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_shared_weights.py:134]: nit: this is usually recommended to be at the start of the loop. gives the best performance and avoids errors.

File: tests/nn/data_parallel/test_fsdp_shared_weights.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            assert sharing is None or sharing == "share_none"
+
+        if with_fsdp:
+            # Shared layers much be un-flatten.
+            self.l1 = FSDP(self.l1, flatten_parameters=False)
+            self.l2 = FSDP(self.l2, flatten_parameters=inner_flat)
+            self.l3 = FSDP(self.l3, flatten_parameters=False)
+
+            if sharing in ["share_only_weights"]:
+                self.l3.append_shared_param(self.l1.module.weight)
```
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_shared_weights.py:49]: Is this a change users will have to make in their models to use this feature?

**Comment by min-xu-ai** [comment]
> LGTM. Can we revert/keep the code simpler if this doesn't end up being useful?

Thanks for the review. Yes definitely.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:441]: This is due to autograd’s hook firing order

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1405]: This is set in the kernel, which is the next PR

File: tests/nn/data_parallel/test_fsdp_shared_weights.py (hunk 3 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_shared_weights.py:134]: I think at the end is in fact slightly better since it reduces the memory immediately after they are used. Where is the source on beginning is better?

File: tests/nn/data_parallel/test_fsdp_shared_weights.py (hunk 4 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_shared_weights.py:49]: Yes

**Comment by min-xu-ai** [comment]
Merged this so that my next PR can properly be based on this. Always OK to address more comments of course.
